ext/tiny_tds/client.c

```diff
@@ -165,6 +171,29 @@ int tinytds_msg_handler(DBPROCESS *dbproc, DBINT msgno, int msgstate, int severi
   return 0;
 }
 
+/*
+Used by dbsetinterrupt -
+This gets called periodically while waiting on a read from the server
+Right now, we only care about cases where a read from the server is
+taking longer than the specified timeout and dbcancel is not working.
+In these cases we decide that we actually want to handle the interrupt
+*/
+static int check_interrupt(void *ptr) {
+  GET_CLIENT_USERDATA((DBPROCESS *)ptr);
+  return userdata->timing_out;
+}
+
+/*
+Used by dbsetinterrupt -
+This gets called if check_interrupt returns TRUE.
+Right now, this is only used in cases where a read from the server is
+taking longer than the specified timeout and dbcancel is not working.
+Return INT_CANCEL to abort the current command batch.
+*/
+static int handle_interrupt(void *ptr) {
+  return INT_CANCEL;
```

Comment:
1. Is it possible for `userdata->timing_out` to go false _after_ `check_interrupt` but _before_ `handle_interrupt`? I wonder if we should do something here like `if (userdata->timing_out) then return INT_CANCEL` to avoid that possibility, and then also maybe set up a pattern for any future interrupt handler logic.
2. Who is responsible for ultimately closing the connection if this is a permanent network failure? Eg, what is the outcome of `#dead?` and `#closed?` after this handler returns? https://github.com/rails-sqlserver/tiny_tds/blob/d8cb94e1033d124dfdd5a08780a530fd8c995101/ext/tiny_tds/client.c#L227-L235

Reply:
1. Great point. This seems like a possibility (though very slim) and I love your suggestion. I'll go ahead and add that check for safety.
2. I'll check up on that and report back. I had assumed both would return `false`.

Reply:
Tested for both this PR and `2.1.3`. Both had the same result:

Timeouts with a working connection:
```
Client dead?: false
Client closed?: false
```

Network timeouts:
```
Client dead?: true
Client closed?: false
```

So it is currently the callers responsibility to close the connection.